### DISCUSSION TITLE
What are the ways to retrieve an element?

### DISCUSSION THREAD

**Discussion by ed2050** [zauberzeug/nicegui#1913] [Q&A] [answered]
### Question

What are the ways to retrieve a nicegui element in your code?  I've read the docs and examples, they don't seem to cover this point, at least not in an obvious place.  I'm struggling with the best way to structure my apps.

# Scenarios
Say we have a nicegui app.  Sometimes we need to access elements from a different context.  For instance:
- a control (button/slider/checkbox/etc) in one page affects which options are shown / enabled in another page
- text in one input affects labels displayed in other parts of the program (e.g. Username field)
- input validation that covers multiple fields (e.g. a weighted average where sum of weights must equal 1)

I don't have time to write a code example now.  Hope the issue is clear.

# Available Options
Here are the options I've seen so far in examples.  However I haven't found any discussion about them: what are the tradeoffs, which approach is best in which situations, etc.
1. Store elements in global variables
2. [Bind variables](https://nicegui.io/documentation/bindings) to controls
3. [Query elements](https://nicegui.io/documentation/query)? only seems to work for retrieving DOM nodes and setting styles / classes / properties, not retrieving nicegui objects and altering values

# Questions
1. Are there any other approaches besides those mentioned above?
2. Are there any docs discussing how to use each approach, tradeoffs, etc?
3. Is there a more general way to retrieve a nicegui object than `ui.query`?  If I want to retrieve the value from a text input element with id=foo, I don't think ui.query will work.

# Discussion
In some situations, option 3 (query) seems like the best approach if available.  But it doesn't seem possible for many use cases, only very simple changes to styles / classes / properties.  If control's value is needed, ui.query doesn't help.

Option 1 (storing global references to elements) requires tracking lots of individual references.  Imagine an app with dozens or hundreds of different controls which can be altered (disabled, value change, color change, etc) by a change to another control.  It becomes a lot of state to manage.

Option 2 (binding) seems to have similar limitations.  First, you need to store the bound values as global references (or at least accessible via a global reference, like a global dict of all values).  Second, it can be harder to update.  I.e. if you bind your Table element's data to the object `bounddict`, then you can't just replace the dict with a new one:
- `bounddict = newdict` changes the object reference, Table doesn't see new data
- instead have to do `bounddict.clear () ; bounddict.update (newdict)` to preserve the original bounddict object which Table refers to.

Binding can work for values, but it feels fragile to me.  You can forget to do the update properly.  And you still have to manage all the bound objects.  Same variable proliferation as approach #1.

Binding doesn't seem to work if you need to change anything other than value or visibility.  If I need to alter the color of the control, what would I bind to?

Querying for elements of a certain type feels more robust to me, as it scales more easily.  Something like:
```py
buttons = ui.findall (ui.button, { classes = 'foo' })  # get all buttons with given properties
for button in buttons : 
   do something ...

inputs = ui.findall (ui.input, parent = { 'type' : ui.card, 'classes' : 'settings' })  # get all input elements on a ui.card with class 'settings'
for input in inputs :  do something...
```
But nothing that general exists as far as I can tell.  Just the limited scope of ui.query.

What's the best way to manage elements across contexts?  Different scopes, different modules, etc.

**Comment by rodja** [comment, marked as answer]
I would say the main current approach is storing elements in variables. These must not be global but can have local scope or be properties of objects. Binding is not a way to access elements but a way to modify them through a model. And `ui.query` is not for finding NiceGUI elements but to get access to objects in the DOM.

> Imagine an app with dozens or hundreds of different controls which can be altered (disabled, value change, color change, etc) by a change to another control. It becomes a lot of state to manage.

I still would argue that you should hold references to these objects and access them properly through the object hierarchy. Python is very powerful in grouping and encapsulating. And this would make the dependency explicit. If one part of the page needs to disable a button somewhere else it's better to make that explicit. All that said, finding and working elements through a query language is appealing for other reasons like reduction of code repetition for styling, working with super dynamic UIs, debugging and prototyping. Thats why I've build upon your `ui.findall` idea in https://github.com/zauberzeug/nicegui/discussions/1923#discussioncomment-7419532. Let us continue over there and close this question as answered in regards to the current state as implemented in 1.4.1. Do you agree?

**Comment by ed2050** [comment]
That's totally fair as an answer to my original question.  Will close now.

Object heirarchies (OH) are fine if that's what you prefer.  I don't want to change how others like to work.  We agree that OH doesn't cover all situations, and a query-based approach is useful in some  situations.  That's good common ground to move forward.

The only minor point I'll make is that object heirarchies aren't necesassily the best "default" approach.  Yes they have the advantages you mention.  They also have downsides.  Namely, duplicating the tree structure of your elements.

Say you store references to objects.  I have a master object to manage references - call it root.  I add elements like this:
```
root.tab = ui.tab ('Search')
with root.tab :
    root.button = ui.button ('Search')
```

As my app grows larger, I have two choices:
1. Give each reference a unique name.  root.tab becomes root.searchtab to distinguish from other tabs.  root.button becomes root.searchtab_searchbutton to distinguish from other search buttons on other tabs.  This quickly becomes unwieldy.
2. Replicate the tree structure of your elements.  root.tab is no longer just a reference, but a node with a reference and children.  So I can do:
```
root.tab = ui.tab ('Search')
with root.tab :
    root.tab.button = ui.button ('Search')
```
(ignoring that you may need extra syntax like root.tab.getelement() to distinguish between accessing the reference and accessing children.  clever underscore methods may avoid this)
 
The problem is you're now maintaining two copies of the object heirarchy: one in nicegui objects, and one in the root container.  What happens when you move an object in the nicegui heirachy?  What happens when you replace a nicegui object with a new object?  The references in root are now out of date and don't reflect your app structure.  It takes extra work to keep them in sync.

To me, this duplication is error prone and should be avoided.  That's why I prefer searching the real element tree over maintaining my own references.  Autocomplete will look _correct_: root does have a reference root.tab.button.  But is that button the same object as my nicegui element tree? 🤷 

Either approach can work.  I like the tradeoffs of query over the tradeoffs of object heirarchies.  So I would disagree that "object heirarchies are better for simple apps".  It depends what style you prefer.

Thanks @rodja! 😄

**Reply by rodja** [reply]
> So I would disagree that "object heirarchies are better for simple apps". It depends what style you prefer.

I would argue that especially in larger applications a good object hierarchy is better to maintain than querying an object graph.
In addition to your two options we often use inheritance to modularize complex user interfaces:

```py
class search_tab(ui.tab_panel):

    def __init__(self):
        super().__init__()
        with self:
            self.button = ui.button('Search')
```

Now after instantiating the `search_tab` you can access the button within via `.button`. I've added a simple version of this to the moudlarization example in PR https://github.com/zauberzeug/nicegui/pull/1929. This has the advantage of not repeating the hierarchy and provides encapsulation to ensure dependencies are visible. The same technique also works without deriving from a UI element. See [automation_controls](https://github.com/zauberzeug/rosys/blob/35b702d986d7308560550f10b0465be8399c6dd0/examples/hello_bot/main.py#L34) for an example which is implemented in [RoSys](https://github.com/zauberzeug/rosys/blob/35b702d986d7308560550f10b0465be8399c6dd0/rosys/automation/automation_controls_.py#L7-L36).

**Reply by ed2050** [reply]
Ok that makes more sense.  I still prefer loosely coupled components, but I see why you prefer your way.

**Comment by rodja** [comment]
We still think that keeping references is better most of the time. But there are corner cases which could benefit from our new `ElementFilter` and a `.mark(...)` method to search for elements in the Python hierarchy: https://nicegui.io/documentation/element_filter

**Reply by ed2050** [reply]
Looks great.  This also makes nicegui more accessible to web devs who program with DOM paradigm.  You're aiming nicegui at broader audience, which is good.  But bringing existing devs to the table builds even more support. 👍
